tools/clang/lib/CodeGen/CGHLSLMSFinishCodeGen.cpp

```diff
@@ -1510,6 +1514,45 @@ Value *EvalBinaryIntrinsic(Constant *cV0, Constant *cV1,
   return Result;
 }
 
+Value *EvalTernaryIntrinsic(Constant *cV0, Constant *cV1, Constant *cV2,
+                             FloatTernaryEvalFuncType floatEvalFunc,
+                             DoubleTernaryEvalFuncType doubleEvalFunc,
+                             IntTernaryEvalFuncType intEvalFunc) {
+  llvm::Type *Ty = cV0->getType();
+  Value *Result = nullptr;
+  if (Ty->isDoubleTy()) {
+    ConstantFP *fpV0 = cast<ConstantFP>(cV0);
+    ConstantFP *fpV1 = cast<ConstantFP>(cV1);
+    ConstantFP *fpV2 = cast<ConstantFP>(cV2);
+    double dV0 = fpV0->getValueAPF().convertToDouble();
+    double dV1 = fpV1->getValueAPF().convertToDouble();
+    double dV2 = fpV2->getValueAPF().convertToDouble();
+    Value *dResult = ConstantFP::get(Ty, doubleEvalFunc(dV0, dV1, dV2));
+    Result = dResult;
+  } else if (Ty->isFloatTy()) {
+    ConstantFP *fpV0 = cast<ConstantFP>(cV0);
+    ConstantFP *fpV1 = cast<ConstantFP>(cV1);
+    ConstantFP *fpV2 = cast<ConstantFP>(cV2);
+    float fV0 = fpV0->getValueAPF().convertToFloat();
+    float fV1 = fpV1->getValueAPF().convertToFloat();
+    float fV2 = fpV2->getValueAPF().convertToFloat();
+    Value *dResult = ConstantFP::get(Ty, floatEvalFunc(fV0, fV1, fV2));
+    Result = dResult;
```

Comment:
It's not related to this change, but I noticed that we don't appear to handle `float16_t` component type and also matrix type. We seem to have code in `CGMSHLSLRuntime::EmitHLSLBuiltinCallExpr` which does wholesale exclusion of `HalfTy`. Maybe this is something we want to change in the future? 

```
else if (operand->getType()->isHalfTy()) {
   // Not support half Eval yet.
   allOperandImm = false;
   break;
}
```

Reply:
True.  However, for half, I don't think it matters much.  The main reason these need to be evaluated here is because of the special literal int/float types. otherwise we end up generating double or i64 ops, which may survive to the end with /Od (and the user didn't ask to use the optional double or 64-bit features).

The only way we get half here is if the user actually used half types for immediate values, in which case, the worst we do is not optimize immediate half ops up-front when using /Od.  But the user shouldn't be surprised by the appearance of halfs when they explicitly asked for them.

Similarly for matrix, normally, the user must specify an explicit component type somewhere to construct the matrix type (you can't actually invoke the literal int/float types in HLSL), so they shouldn't be surprised by the operation type selected.  Plus there's a bunch of different code that would have to be used to deal with the matrix operations, that HLMatrixLowerPass currently handles.

However, I realize that we could run into a new issue with templates.  Normally a user could not construct a matrix type with a literal component type (because there's no usable HLSL symbol corresponding to these types), but with templates you could in theory construct a matrix with a literal float/int when the template instantiation has a typename based on a literal float/int.  I actually think this is problematic even more generally, so perhaps we need to require a concrete types for template typename resolution and instantiation to avoid the general issue.  Otherwise you would be able to start to use literal float/int for all sorts of non-literal user code and data structures!

Reply:
Sounds good. thanks for the detailed explanation!